src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java

```diff
@@ -0,0 +1,249 @@
+
+    /** The maximum size of the cache in bytes. */
+    private final int mMaxCacheSizeInBytes;
+
+    /** Default maximum disk usage in bytes. */
+    private static final int DEFAULT_DISK_USAGE_BYTES = 5 * 1024 * 1024;
+
+    /** High water mark percentage for the cache */
+    @VisibleForTesting static final float HYSTERESIS_FACTOR = 0.9f;
+
+    /** Magic number for current version of cache file format. */
+    private static final int CACHE_MAGIC = 0x20150306;
+
+    /**
+     * Constructs an instance of the DiskBasedAsyncCache at the specified directory.
+     *
+     * @param rootDirectory The root directory of the cache.
+     * @param maxCacheSizeInBytes The maximum size of the cache in bytes. Note that the cache may
+     *     briefly exceed this size on disk when writing a new entry that pushes it over the limit
+     *     until the ensuing pruning completes.
+     */
+    public DiskBasedAsyncCache(final File rootDirectory, int maxCacheSizeInBytes) {
+        mRootDirectorySupplier =
+                new DiskBasedAsyncCache.FileSupplier() {
+                    @Override
+                    public File get() {
+                        return rootDirectory;
+                    }
+                };
+        mMaxCacheSizeInBytes = maxCacheSizeInBytes;
+    }
+
+    /** Clears the cache. Deletes all cached files from disk. */
+    @Override
+    public synchronized void clear() {
+        // TODO (sphill99): Implement
+    }
+
+    /** Returns the cache entry with the specified key if it exists, null otherwise. */
+    @Override
+    public void get(String key, OnGetCompleteCallback callback) {
+        final OnGetCompleteCallback cb = callback;
+        final DiskBasedAsyncCache.CacheHeader entry = mEntries.get(key);
+        File file = getFileForKey(key);
+        Path path = Paths.get(file.getPath());
+        try {
+            AsynchronousFileChannel afc =
+                    AsynchronousFileChannel.open(path, StandardOpenOption.READ);
+            ByteBuffer buffer = ByteBuffer.allocate((int) file.length());
+            afc.read(
+                    buffer,
+                    0,
+                    buffer,
+                    new CompletionHandler<Integer, ByteBuffer>() {
+                        @Override
+                        public void completed(Integer result, ByteBuffer attachment) {
+                            if (attachment.hasArray()) {
+                                int offset = attachment.arrayOffset();
+                                byte[] data = attachment.array();
+                                cb.onGetComplete(entry.toCacheEntry(data));
+                            }
+                        }
+
+                        @Override
+                        public void failed(Throwable exc, ByteBuffer attachment) {
+                            cb.onGetComplete(null);
+                        }
+                    });
+        } catch (Exception e) {
```

Comment:
The `DiskBasedCache` only catchtes `IOException`s. Is there a reason why we want to catch any `Exception`s in here?

Reply:
Oh no, I think I just put that there when I first wrote the try/catch, and never changed it. I'll fix that now.